### ISSUE TITLE
Eligibility criteria for routine vaccination

### ISSUE THREAD

**Issue by wagathu** [starsimhub/starsim#640]
Hi, how do we define the eligibility of routine vaccination (ss.routine_vx()) to be a function of age (sim.people.age)? I tried 
```
# Defining eligibility
def eligibility_function(sim):
    ages = sim.people.age 
    is_eligible = (ages >= 9) & (ages <= 12)
    return sim.people.auids[is_eligible]

my_intervention1 = ss.routine_vx(
    start_year=2020,
    prob=0.95,         
    product=mcv1,       
    eligibility=eligibility_function  
)

But still people with ages greater than 12 or less than 9 months get vaccinated

```

**Comment by cliffckerr** [comment]
Thank you for finding this, it's a bug! Working on a fix, but for now, if you move the eligibility outside the definition, it works:
```py
import starsim as ss

# Defining eligibility
def eligibility_function(sim):
    ages = sim.people.age 
    is_eligible = (ages >= 9) & (ages <= 12)
    print(f'Checking eligibility: of {len(ages)} people, {is_eligible.sum()} are eligible')
    return sim.people.auids[is_eligible]

# Define a custom "SIS" vaccine
class sis_vaccine(ss.Vx):
    def __init__(self, pars=None, *args, **kwargs):
        super().__init__()
        self.default_pars(
            efficacy = 0.9,
        )
        self.update_pars(pars, **kwargs)
        return

    def administer(self, people, uids):
        ages = people.age[uids]
        min_age = ages.min()
        max_age = ages.max()
        rel_sus = people.sis.rel_sus
        rel_sus[uids] *= 1-self.pars.efficacy
        print(f'Administering {len(uids)} doses for people aged {min_age:n} - {max_age:n}; rel_sus={rel_sus.mean():n}')
        return

# Create the product
mcv1 = sis_vaccine()

# Create the intervention
my_intervention1 = ss.routine_vx(
    start_year=2020,
    prob=0.9,         
    product=mcv1,       
)
my_intervention1.eligibility = eligibility_function # Move this outside of the parameters -- sorry!!

# Run the simulation
sim = ss.Sim(diseases='sis', networks='random', interventions=my_intervention1)
sim.run()
sim.plot()
```
(The effect of vaccination is too small to see in the plot, but it's there.)

**Comment by cliffckerr** [comment]
OK, I found the fix -- we can just use `super()`:
```py
import sciris as sc

class Intervention(sc.prettyobj):
    def __init__(self, eligibility=None, *args, **kwargs):
        self.args = args
        self.kwargs = kwargs
        self.eligibility = eligibility
        return

class BaseScreening(Intervention):
    def __init__(self, product=None, prob=None, eligibility=None, **kwargs):
        super().__init__(**kwargs)
        self.prob = prob
        self.eligibility = eligibility
        return
        
class RoutineDelivery(Intervention):
    def __init__(self, years=None, start_year=None, end_year=None, prob=None, annual_prob=True, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.years = years
        self.start_year = start_year
        self.end_year = end_year
        self.prob = prob
        self.annual_prob = annual_prob  # Determines whether the probability is annual or per timestep
        return

# Current approach -- does not work
class routine_screening(BaseScreening, RoutineDelivery):
    def __init__(self, product='a', prob='b', eligibility='c',
                 years='d', start_year='e', end_year='f', **kwargs):
        BaseScreening.__init__(self, product=product, eligibility=eligibility, **kwargs)
        RoutineDelivery.__init__(self, prob=prob, start_year=start_year, end_year=end_year, years=years)
        return
    
# Correct approach -- just use super()
class routine_screening2(BaseScreening, RoutineDelivery):
    def __init__(self, product='a', prob='b', eligibility='c',
                 years='d', start_year='e', end_year='f', **kwargs):
        super().__init__(product=product, eligibility=eligibility, prob=prob, 
                         start_year=start_year, end_year=end_year, years=years, **kwargs)
        return
    
# Test
rs = routine_screening()
rs2 = routine_screening2()
print(rs)
print(rs2)
assert rs2.eligibility == 'c'
```

Tomorrow will release a v1.0.3 of Starsim with this fix. Thanks again for finding it!

**Comment by cliffckerr** [comment]
Thanks @wagathu! This bug has been fixed in the latest version of Starsim (1.0.3). It will also be available in the new Starsim release (2.0). (Closed by #651.)

**Comment by wagathu** [comment]
Thank you for this.
